src/libstd/thread/mod.rs

```diff
@@ -1058,23 +1066,18 @@ impl Thread {
     /// [park]: fn.park.html
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn unpark(&self) {
-        loop {
-            match self.inner.state.compare_exchange(EMPTY, NOTIFIED, SeqCst, SeqCst) {
-                Ok(_) => return, // no one was waiting
-                Err(NOTIFIED) => return, // already unparked
-                Err(PARKED) => {} // gotta go wake someone up
-                _ => panic!("inconsistent state in unpark"),
-            }
-
-            // Coordinate wakeup through the mutex and a condvar notification
-            let _lock = self.inner.lock.lock().unwrap();
-            match self.inner.state.compare_exchange(PARKED, NOTIFIED, SeqCst, SeqCst) {
-                Ok(_) => return self.inner.cvar.notify_one(),
-                Err(NOTIFIED) => return, // a different thread unparked
-                Err(EMPTY) => {} // parked thread went away, try again
-                _ => panic!("inconsistent state in unpark"),
-            }
+        // We must unconditionally write NOTIFIED here to
+        // synchronize with a read in `park`.
```

Comment:
I think I don't quite fully understand this comment and why the loop isn't needed any more. Can you expand a bit on why this unconditional write is needed? Also is this a case like above where we're not actually synchronizing with `park` at all but rather we're synchronizing with reads after `park` returns?

Reply:
In `unpark()`, we want to do some kind of `Release` write in order to "publish" the current thread's memory, even if the thread has already been notified.

Later on, a call to `park()` will do an `Acquire` load and be able to see that "published" memory. If we skipped the `Release` write in `park()` then the `Acquire` load wouldn't be able to see it.

Put differently, whenever a `park()` clears the `NOTIFIED` flag, we want that to happen-before (`Acquire`-`Release`) every `unpark()` that set the `NOTIFIED` flag. The point is that `unpark()` must set the flag to `NOTIFIED` even if it was already `NOTIFIED`. :)

Does that make any sense?

Reply:
The loop isn't needed anymore it is just doing a swap instead of a CAS, so there is no failure/retry case anymore. We are doing a swap be because each `unpark` must write a value that a `park` reads to ensure the happens-before relationship of any writes before `unpark` and any writes after `park` returns.

Reply:
Ok thanks for the info! This was also suspect to me because the mutex isn't protecting anything in this code path, but I think it's ok because the atomic comparison above is done inside of the mutex.